api-server/src/server/boot/user.js

```diff
@@ -468,6 +474,42 @@ function postResetProgress(req, res, next) {
   );
 }
 
+async function postResetProjectProgress(req, res, next) {
+  try {
+    const { user } = req;
+    const { dashedBlockName } = req.params;
+
+    const toResetChallengeIds = getChallengeIdsForBlock(dashedBlockName);
+
+    const originalCount = user.completedChallenges.length;
+    const updatedCompletedChallenges = user.completedChallenges.filter(
+      challenge => !toResetChallengeIds.includes(challenge.id)
+    );
+
+    if (originalCount === updatedCompletedChallenges.length) {
+      return res.status(200).json({
+        message: `No challenges were reset because the user had not completed any challenges in ${dashedBlockName}`
+      });
+    }
+
+    await user.updateAttributes({
+      completedChallenges: updatedCompletedChallenges
+    });
```

Comment:
This is unnecessarily expensive. We should try use Mongo's operators like:

```js
db.collection('user').updateOne(
      {},
      { $pull: { completedChallenges: { id: { $in: ids } } } }
    );
```

NOTE: Do not use the above. We should investigate the "Loopback" way of doing such a query.

Reply:
Thank you for the review, @ShaunSHamilton. As suggested, I've updated the implementation to use MongoDB's operators efficiently within LoopBack's framework. The function now leverages the `$pull` operator to modify the `completedChallenges` array directly in the database. This approach not only improves performance by reducing the overhead of manual array manipulation in the application layer but also ensures that updates are only made when there are changes to be committed, thus minimizing unnecessary database writes.

These enhancements are part of the latest commit: [34a620b84da16eeb40496456a55f483f32f2e979].

If there are any further improvements or adjustments to consider, please let me know.